mono/metadata/threads.c

```diff
@@ -1534,9 +1557,12 @@ ves_icall_System_Threading_Thread_Sleep_internal(gint32 ms)
 		mono_thread_clr_state (thread, ThreadState_WaitSleepJoin);
 
 		if (alerted) {
-			MonoException* exc = mono_thread_execute_interruption ();
-			if (exc) {
-				mono_set_pending_exception (exc);
+			MonoExceptionHandle exc = mono_thread_execute_interruption_handle (error);
+			mono_error_assert_ok (error);
+			if (exc && !MONO_HANDLE_IS_NULL (exc)) {
+				mono_set_pending_exception_handle (exc);
```

Comment:
Instead of adding a variant of `mono_set_pending_exception` I think a better approach is to make a version of `mono_thread_execute_interruption` that sets a `MonoError` if there was an interruption.  We want fewer uses of `mono_set_pending_exception` in the runtime, not more.

Reply:
This code has proven very sensitive and fails upon making correct looking changes.

Reply:
I don't think this made things worse in this regard either but I can look again.

Reply:
The exception is not always set to pending so that will require a little more refactoring.

Reply:
This somewhat confuses an exception for interruption vs. an exception creating said exception.
The first is expected sometimes, the second is low memory and not expected -- but we do thread the error parameter every where and this is the main reason it can be set, even though we ignore low memory sometimes.

Reply:
This will maybe mess with lifetime. Which maybe is the problem I had changing this area in the first place. i.e. other than returning the exception, we also set thread->abort_exc and/or make it pending.

Reply:
I suggest we leave this asis for this PR. The code is a bit messy and I made a pretty minimal mechanical change.